log.js

```diff
@@ -0,0 +1,106 @@
+const { writeFileSync, readFileSync } = require("fs");
+const { stringify, parse } = require("yaml");
+
+// log file
+const logFile = "log.yaml";
+
+// load previous runs of bot from log file
+function loadLog() {
+  try {
+    const file = readFileSync(logFile, { encoding: "utf8" });
+    const data = parse(file);
+    return data;
+  } catch (error) {
+    return null;
+  }
+}
+
+// save this run of bot to log file
+function saveLog({ preprint, comment, tweets }, log = []) {
+  try {
+    log.push({
+      preprint: cleanPreprint(preprint),
+      comment: cleanComment(comment),
+      tweets: cleanTweets(tweets),
+    });
+
+    writeFileSync(logFile, stringify(log), () => null, { encoding: "utf8" });
+    return true;
+  } catch (error) {
+    return null;
+  }
+}
+
+// keep only useful props from preprint data
+const cleanPreprint = (preprint) => {
+  for (const authorIndex of Object.keys(preprint.authors)) {
+    delete preprint?.authors[authorIndex]?.id;
+    delete preprint?.authors[authorIndex]?.institution;
+  }
+  delete preprint?.ranks;
+  delete preprint?.publication;
+
+  return preprint;
+};
+
+// keep only useful props from comment data
+const cleanComment = (comment) => {
+  delete comment?.editableUntil;
+  delete comment?.message;
+  delete comment?.author?.disable3rdPartyTrackers;
+  delete comment?.author?.joinedAt;
+  delete comment?.author?.isPrivate;
+  delete comment?.author?.signedUrl;
+  delete comment?.author?.isPrimary;
+  delete comment?.author?.isAnonymous;
+  delete comment?.author?.avatar;
+  delete comment?.media;
+  delete comment?.isSpam;
+  delete comment?.isDeletedByAuthor;
+  delete comment?.isHighlighted;
+  delete comment?.parent;
+  delete comment?.isDeleted;
+  delete comment?.isFlagged;
+  delete comment?.isAtFlagLimit;
+  delete comment?.canVote;
+  delete comment?.moderationLabels;
+  delete comment?.isEdited;
+  delete comment?.sb;
+
+  return comment;
+};
+
+// keep only useful props from tweets data
+const cleanTweets = (tweets) => {
```

Comment:
Do you think it's better to remove keys (like this) or would it be better to create a new list and only include selected keys? I am imagining that they might add properties in the future (maybe that never happens).

Reply:
I started doing it as a white list rather than a black list, but I found this to be a bit cleaner. I want to maintain the structure of the original data from the APIs, and there's a lot of nested fields. Recreating them made the code less readable imo. More indentation and destructuring. This looks more repetitive, but it's more flat and simple.

Also wanted to err on the side of keeping things rather than deleting them, so opting out seemed the right way to do it.

But I'm not 100% attached to this. I could show you what it looks like the other way.